arch: arc: clear MPU regions before init it to fix tests exception and timeout

YuguoWH opened the pull request:
EMSK boards can't be reset between tests due to board hardware design. 
But MPU v3 assume that all the MPU registers has been reset to zero. 
During sanitycheck tests, MPU configs of previous test could pass to following tests and cause exceptions. 
This PR fixes this issue by adding MPU region reset. 
The issue was only found on MPU v3, so this fix does not apply to MPU v2.

Fixes #28493

abrodkin commented on arch/arc/core/mpu/arc_mpu_v3_internal.h:
```diff
@@ -845,6 +845,10 @@ static int arc_mpu_init(const struct device *arg)
 	/* Disable MPU */
 	arc_core_mpu_disable();
 
+	/* clear all the regions first */
```
Please add your perfectly correct rationale in comments here so every reader of that code understands why do we do that. Though given we don't limit that fix-up to a particular board (which will add an extra `ifdef` which is not good) explain it in a more generic way (i.e. w/o mention of EMSK).

YuguoWH replied:
Hi @abrodkin , thank you for the advice. Modified.

ruuddw reviewed (changes requested):
The idea here is correct, but I think the solution can be improved:

mpu_disable() is performed and the existing init logic programs all regions -> not obvious why the initial zeroing is required/
The problem is that mpu_disable() doesn't actually disable the MPU (v3 cannot be disabled), but it is emulated by programming the default region. This ignores that there can be other regions programmed with access restrictions.
Simply clearing all regions in mpu_disable() isn't a solution, as mpu_disable should not destroy the regions state.
arc_mpu_init() is called reasonably early, but if the intent is to put the MPU in 'reset' state, I think it is better to do this in the very early reset code.

So, please either:

- move the MPU region clearing + programming the default region to allow all access to early init code (with a comment that it is for emulating some part of a HW reset when doing a soft 'reset')

- or add the region clearing to the mpu_disable() -> but this will require more changes: remove/change other calls to mpu_disable() that expect disable but with state of the regions kept.

Thanks,

Ruud.

abrodkin commented on arch/arc/core/reset.S:
```diff
@@ -121,6 +121,35 @@ invalidate_dcache:
+#if defined(CONFIG_ARC_CUSTOM_INIT)
+#if CONFIG_ARC_MPU_VER == 3
+	/* Set default reset value to _ARC_V2_MPU_EN */
+	mov_s r1, 0x400181C0
```
Please don't use literals like that. Create a definition right here with meaningful name & comments.

ruuddw commented on arch/arc/core/reset.S:
```diff
@@ -121,6 +121,35 @@ invalidate_dcache:
+/*
+ * ARC Custom Initialization
+ * Force Reset MPU (v3) registers
+ * For some test scenarios, this is needed to apply additional configurations
```
Suggest to change/clarify: "For scenarios where the board hardware is not reset in between tests, some settings need to be reset in software"

abrodkin replied:
We're trying to mimic real "cold reset" with additional initialization.

abrodkin commented on arch/arc/core/mpu/arc_mpu_v2_internal.h:
```diff
@@ -16,10 +16,6 @@
 #define AUX_MPU_RDP_ATTR_MASK (0x1FC)
 #define AUX_MPU_RDP_SIZE_MASK (0xE03)
 
-#define _ARC_V2_MPU_EN   (0x409)
```
Cannot we move that non functional change to the separate commit (even if you prefer to include it in this PR)?

abrodkin commented on arch/arc/core/mpu/arc_mpu_v3_internal.h:
```diff
@@ -12,16 +12,6 @@
 
 #define AUX_MPU_RPER_ATTR_MASK (0x1FF)
 
-#define _ARC_V2_MPU_EN   (0x409)
```
Ditto.

ruuddw commented on include/arch/arc/v2/aux_regs.h:
```diff
@@ -110,6 +110,17 @@
 #define _ARC_V2_FPU_DPFP2L 0x304
 #define _ARC_V2_FPU_DPFP2H 0x305
+#define _ARC_V2_MPU_EN 0x409
+#if CONFIG_ARC_MPU_VER == 2
```
no need for conditional compilations here, the values all are different and we don't have conditions for other HW features like e.g. FPU.

ruuddw reviewed (approved):
Just a few minor suggestions for improving.

YuguoWH committed:
arch: arc: unify different versions of MPU registers

Previously MPU registers macros are only defined within its own header
files and could not be used by other part of program. This commit unify
them together.

Signed-off-by: Yuguo Zou <yuguo.zou@synopsys.com>

YuguoWH committed:
arch: arc: Restore MPU registers to its initial states between tests

EMSK boards can't be reset between tests due to hardware configures.
MPU v3 configs in previous test could cause exceptions in the following
tests. This commit fixes this issue by restoring MPU registers initial
states at early init stage.

Signed-off-by: Yuguo Zou <yuguo.zou@synopsys.com>

YuguoWH commented on include/arch/arc/v2/aux_regs.h:
```diff
@@ -110,6 +110,17 @@
+#define _ARC_V2_MPU_PROBE 0x44C
+
+/* Default reset value for _ARC_V2_MPU_EN register */
+#define ARC_MPU_EN_RESET_VALUE 0x400181C0
```
Have some doubt here: I wish to put this in arc_mpu.h, but the header file is not included in assembly code.

abrodkin replied:
There's no need to move that default value which is used only one in 1 place into generic header. And that's what I meant in my initial commit - move that definition right in the place where it is used. It's OK to have it done that way:
```c
...
// This is very important definition used for XXX
#define MY_IMPORTNAT_DEFINITION 0x12345678
    super_secret_password = MY_IMPORTNAT_DEFINITION;
...
```

abrodkin commented on arch/arc/Kconfig:
```diff
@@ -328,4 +328,11 @@ config ARC_EXCEPTION_DEBUG
 	  and parameters, at a cost of code/data size for the human-readable
 	  strings.
 
+config ARC_CUSTOM_INIT
```
I really like that definition & help message!

abrodkin commented on arch/arc/core/reset.S:
```diff
@@ -121,6 +121,36 @@ invalidate_dcache:
+/*
+ * ARC Custom Initialization
+ * For scenarios where board hardware is not reset between tests,
+ * Some settings need to be reset by software.
```
I'd prefer to not use a word "reset" because we don't reset anything here in a typical sense.. i.e. HW is not being re-initialized internally. Instead we return its settings to the default initial state trying to substitute normal HW reset sequence.

abrodkin commented on arch/arc/core/reset.S:
```diff
@@ -121,6 +121,36 @@ invalidate_dcache:
+	/* Software Reset MPU (v3) registers */
+#if CONFIG_ARC_MPU_VER == 3
+	/* Set default reset value to _ARC_V2_MPU_EN */
+	mov_s r1, ARC_MPU_EN_RESET_VALUE
```
As I mentioned move its definition right here this way
```c
/* Default reset value for _ARC_V2_MPU_EN register */
#define ARC_MPU_EN_RESET_VALUE 0x400181C0

    /* Set default reset value to _ARC_V2_MPU_EN */
    mov_s r1, ARC_MPU_EN_RESET_VALUE
    sr r1, [_ARC_V2_MPU_EN]
```

abrodkin commented on arch/arc/core/reset.S:
```diff
@@ -121,6 +121,36 @@ invalidate_dcache:
+	mov_s r1, 0
+	mov_s r2, 0
+	/* Reset all MPU regions by iterating index */
+mpu_software_reset:
```
Here as well that `mpu_software_reset` label is a bit confusing.
Maybe something more functional like `mpu_regions_reset`?

abrodkin reviewed (changes requested):
I'd suggest to rephrase "arch: arc: Add MPU software reset at early init stage" to something w/o mention of "software reset" as we cannot do any real reset with software, instead we just "return MPU settings to defaults".

ruuddw reviewed (approved):
Even better like this, thanks.

ioannisg commented on arch/arc/Kconfig:
```diff
@@ -328,4 +328,11 @@ config ARC_EXCEPTION_DEBUG
 
+config ARC_CUSTOM_INIT
+	bool "Enable custom initialization code"
+	default n
```
Default n is not needed, here (it is default n, anyways), so it should be removed

ioannisg commented:
This looks OK to me.
I just feel that eventually we need to make a decision whether Zephyr should really initialize every ARCH-level hardware once it is coming out of reset or it should keep relying on the parent image, e.g. a bootloader... Right now we are not consistent across architectures and platforms.

abrodkin commented:
> This looks OK to me.
> I just feel that eventually we need to make a decision whether Zephyr should really initialize every ARCH-level hardware once it is coming out of reset or it should keep relying on the parent image, e.g. a bootloader... Right now we are not consistent across architectures and platforms.

So are there any similar hacks done for other boards/arches? I'm asking because IMHO it would be logical to use some unified ifdef for them so that we may:
1. Document it and use similarly across the board
2. Easier remove all that mess if we decide to rely solely on clean out-of-reset state

ioannisg commented:
> So are there any similar hacks done for other boards/arches? I'm asking because IMHO it would be logical to use some unified ifdef for them so that we may:

> Document it and use similarly across the board
> Easier remove all that mess if we decide to rely solely on clean out-of-reset state

Yes, at least Cortex-M is doing similar things, everywhere in the system init code. 

I guess the Kconfig switch should be something like `config ARCH_INIT_HW_AT_BOOT` and should be used explicitly around code that initializes ARCH core hardware to its default values, if it were coming out of warm reset.

ioannisg commented:
@abrodkin if you would like to open a GH ticket for this, I can fill it in with the information about ARM.

YuguoWH commented:
Removed `default n` in Kconfig

abrodkin commented:
@ioannisg see https://github.com/zephyrproject-rtos/zephyr/issues/28803
